### PR TITLE
feat: add functions to support use of timestamp data type

### PR DESCRIPTION
**Pull request by jzaralim**
### Description 
Functions added:
* unix_timestamp(timestamp) - converts timestamp to bigint
* convert_tz - convert timestamps between timezones
* formattimestamp - convert timestamp to string
* parsetimestamp - convert string to timestamp
* from_unixtime - converts bigint to timestamp

### Testing done 
unit tests, qtt

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: docs/developer-guide/ksqldb-reference/scalar-functions.md [hunk 1]
```diff
@@ -1096,6 +1104,52 @@ java.util.TimeZone ID format, for example: "UTC",
...
+
+```sql
+FORMATTIMESTAMP(timestamp, 'yyyy-MM-dd HH:mm:ss.SSS' [, TIMEZONE])
+```
+
+Converts a TIMESTAMP value into the string representation of the timestamp in the given format.
+Single quotes in the timestamp format can be escaped with two successive single quotes, `''`, for
+example: `'yyyy-MM-dd''T''HH:mm:ssX'`.
+
+TIMEZONE is an optional parameter and it is a java.util.TimeZone ID format, for example: "UTC",
```
Comment (JimGalasyn) [inline docs/developer-guide/ksqldb-reference/scalar-functions.md:1117]:
```suggestion
TIMEZONE is an optional parameter and it is a `java.util.TimeZone` ID format, for example: "UTC",
```

**Review by JimGalasyn — APPROVED** [review]
LGTM!

File: docs/developer-guide/ksqldb-reference/scalar-functions.md [hunk 2]
```diff
@@ -1096,6 +1104,52 @@ java.util.TimeZone ID format, for example: "UTC",
...
+format can be escaped with two successive single quotes, `''`, for example: `'yyyy-MM-dd''T''HH:mm:ssX'`.
+
+TIMEZONE is an optional parameter and it is a `java.util.TimeZone` ID format, for example: "UTC",
+"America/Los_Angeles", "PDT", "Europe/London". For more information on timestamp formats, see
+[DateTimeFormatter](https://cnfl.io/java-dtf).
+
+### `CONVERT_TZ`
+
+```sql
+CONVERT_TZ(col1, 'from_timezone' 'to_timezone')
```
Comment (spena) [inline docs/developer-guide/ksqldb-reference/scalar-functions.md:1137]: Comma?

```suggestion
CONVERT_TZ(col1, 'from_timezone', 'to_timezone')
```

File: ksqldb-common/src/main/java/io/confluent/ksql/util/timestamp/StringToTimestampParser.java [hunk 3]
```diff
@@ -40,6 +41,25 @@ public StringToTimestampParser(final String pattern) {
...
+  /**
+   * @param text    the textual representation of the timestamp
+   * @param zoneId  the zoneId to use, if none present in {@code text}
+   *
+   * @return the millis since epoch that {@code text} represents
+   */
+  public Timestamp parseToTimestamp(final String text, final ZoneId zoneId) {
+    return Timestamp.from(parseZoned(text, zoneId).toInstant());
+  }
+
```
Comment (spena) [inline ksqldb-common/src/main/java/io/confluent/ksql/util/timestamp/StringToTimestampParser.java:62]: Are there tests for these two functions? I don't see the `StringToTimestampParserTest`

File: ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/datetime/ConvertTz.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+  @Udf(description = "Converts a TIMESTAMP value from one timezone to another")
+  public Timestamp convertTz(
+      @UdfParameter(
+          description = "the TIMESTAMP value.") final Timestamp timestamp,
+      @UdfParameter(
+          description =  " fromTimeZone is a java.util.TimeZone ID format, for example: \"UTC\","
+              + " \"America/Los_Angeles\", \"PST\", \"Europe/London\"") final String fromTimeZone,
+      @UdfParameter(
+          description =  " toTimeZone is a java.util.TimeZone ID format, for example: \"UTC\","
+              + " \"America/Los_Angeles\", \"PST\", \"Europe/London\"") final String toTimeZone
```
Comment (spena) [inline ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/datetime/ConvertTz.java:45]:
```suggestion
      @UdfParameter(
          description = "The TIMESTAMP value.") final Timestamp timestamp,
      @UdfParameter(
          description =  "The fromTimeZone in a java.util.TimeZone ID format. For example: \"UTC\","
              + " \"America/Los_Angeles\", \"PST\", \"Europe/London\"") final String fromTimeZone,
      @UdfParameter(
          description =  " The toTimeZone in a java.util.TimeZone ID format. For example: \"UTC\","
              + " \"America/Los_Angeles\", \"PST\", \"Europe/London\"") final String toTimeZone
```

File: ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/datetime/FormatTimestamp.java [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+import io.confluent.ksql.function.udf.UdfDescription;
+import io.confluent.ksql.function.udf.UdfParameter;
+import io.confluent.ksql.util.KsqlConstants;
+import java.sql.Timestamp;
+import java.time.ZoneId;
+import java.time.format.DateTimeFormatter;
+import java.util.concurrent.ExecutionException;
+
+@UdfDescription(
+    name = "formattimestamp",
```
Comment (spena) [inline ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/datetime/FormatTimestamp.java:33]:
```suggestion
    name = "format_timestamp",
```

File: ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/datetime/FormatTimestamp.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+          .atZone(ZoneId.systemDefault())
+          .format(formatter);
+    } catch (final ExecutionException | RuntimeException e) {
+      throw new KsqlFunctionException("Failed to format timestamp " + timestamp
+          + " with formatter '" + formatPattern
+          + "': " + e.getMessage(), e);
+    }
+  }
+
+  @Udf(description = "Converts a BIGINT millisecond timestamp value into the"
```
Comment (spena) [inline ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/datetime/FormatTimestamp.java:70]:
```suggestion
  @Udf(description = "Converts a TIMESTAMP value into the"
```

File: ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/datetime/FormatTimestamp.java [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+      final DateTimeFormatter formatter = formatters.get(formatPattern);
+      final ZoneId zoneId = ZoneId.of(timeZone);
+      return timestamp.toInstant()
+          .atZone(zoneId)
+          .format(formatter);
+    } catch (final ExecutionException | RuntimeException e) {
+      throw new KsqlFunctionException("Failed to format timestamp " + timestamp
+          + " at timeZone '" + timeZone + "' with formatter '" + formatPattern
+          + "': " + e.getMessage(), e);
+    }
```
Comment (spena) [inline ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/datetime/FormatTimestamp.java:92]: All this code is similar to the other formatTimestamp except for the timeZone. Would it be good to have a method for this?

File: ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/datetime/ParseTimestamp.java [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+import io.confluent.ksql.function.udf.UdfDescription;
+import io.confluent.ksql.function.udf.UdfParameter;
+import io.confluent.ksql.util.KsqlConstants;
+import io.confluent.ksql.util.timestamp.StringToTimestampParser;
+import java.sql.Timestamp;
+import java.time.ZoneId;
+import java.util.concurrent.ExecutionException;
+
+@UdfDescription(
+    name = "parsetimestamp",
```
Comment (spena) [inline ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/datetime/ParseTimestamp.java:33]:
```suggestion
    name = "parse_timestamp",
```

File: ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/datetime/ParseTimestamp.java [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+)
+public class ParseTimestamp {
+
+  private final LoadingCache<String, StringToTimestampParser> parsers =
+      CacheBuilder.newBuilder()
+          .maximumSize(1000)
+          .build(CacheLoader.from(StringToTimestampParser::new));
+
+  @Udf(description = "Converts a string representation of a date in the given format"
+      + " into the BIGINT value that represents the millisecond timestamp."
```
Comment (spena) [inline ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/datetime/ParseTimestamp.java:47]:
```suggestion
      + " into the TIMESTAMP value that represents the timestamp."
```

File: ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/datetime/ParseTimestamp.java [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+      return timestampParser.parseToTimestamp(formattedTimestamp);
+    } catch (final ExecutionException | RuntimeException e) {
+      throw new KsqlFunctionException("Failed to parse timestamp '" + formattedTimestamp
+          + "' with formatter '" + formatPattern
+          + "': " + e.getMessage(), e);
+    }
+  }
+
+  @Udf(description = "Converts a string representation of a date at the given time zone"
+      + " in the given format into the BIGINT value that represents the millisecond timestamp."
```
Comment (spena) [inline ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/datetime/ParseTimestamp.java:67]:
```suggestion
      + " in the given format into the TIMESTAMP value that represents the timestamp."
```

File: ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/datetime/ParseTimestamp.java [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+              + " \"America/Los_Angeles\", \"PST\", \"Europe/London\"") final String timeZone) {
+    try {
+      final StringToTimestampParser timestampParser = parsers.get(formatPattern);
+      final ZoneId zoneId = ZoneId.of(timeZone);
+      return timestampParser.parseToTimestamp(formattedTimestamp, zoneId);
+    } catch (final ExecutionException | RuntimeException e) {
+      throw new KsqlFunctionException("Failed to parse timestamp '" + formattedTimestamp
+          + "' at timezone '" + timeZone + "' with formatter '" + formatPattern
+          + "': " + e.getMessage(), e);
+    }
```
Comment (spena) [inline ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/datetime/ParseTimestamp.java:87]: This code is similar to the other function. Can we wrap this into one method?

File: ksqldb-functional-tests/src/test/resources/query-validation-tests/stringtimestamp.json [hunk 12]
```diff
@@ -1,11 +1,6 @@
 {
   "comments": [
-    "You can specify multiple statements per test case, i.e., to set up the various streams needed",
-    "for joins etc, but currently only the final topology will be verified. This should be enough",
-    "for most tests as we can simulate the outputs from previous stages into the final stage. If we",
-    "take a modular approach to testing we can still verify that it all works correctly, i.e, if we",
-    "verify the output of a select or aggregate is correct, we can use simulated output to feed into",
-    "a join or another aggregate."
+    "Tests covering the use of the STRINGTOTIMESTAMP and  PARSETIMESTAMP functions."
```
Comment (spena) [inline ksqldb-functional-tests/src/test/resources/query-validation-tests/stringtimestamp.json:3]:
```suggestion
    "Tests covering the use of the STRINGTOTIMESTAMP and  PARSE_TIMESTAMP functions."
```

File: ksqldb-functional-tests/src/test/resources/query-validation-tests/timestamp-to-string.json [hunk 13]
```diff
@@ -1,10 +1,10 @@
 {
   "comments": [
-    "Tests covering the use of the TIMESTAMPTOSTRING function."
+    "Tests covering the use of the TIMESTAMPTOSTRING and  FORMATTIMESTAMP functions."
```
Comment (spena) [inline ksqldb-functional-tests/src/test/resources/query-validation-tests/timestamp-to-string.json:3]:
```suggestion
    "Tests covering the use of the TIMESTAMPTOSTRING and  FORMAT_TIMESTAMP functions."
```

**Review by spena — APPROVED** [review]
LGTM!. thanks @jzaralim

**Pull request merged**